packages/cli/lib/model-discoverer.js

```diff
@@ -0,0 +1,105 @@
+const debug = require('./debug')('model-discoverer');
+const fs = require('fs');
+const path = require('path');
+
+/**
+ * Given a datasource and discovery options,
+ * return a list of objects {table: 'foo', schema: 'bar}
+ */
+async function discoverModelNames(ds, options) {
+  if (!ds.connected) {
+    await new Promise(resolve => {
+      ds.on('connected', resolve);
+    });
+  }
+  return await ds.discoverModelDefinitions(options);
+}
+
+/**
+ * Returns the entire model definition for a model
+ */
+async function discoverSingleModel(ds, options, name) {
+  const schema = await ds.discoverSchema(name, options);
+  schema.settings = schema.options;
+  return schema;
+}
+
+/**
+ * Loads a DataSource from a file
+ * If the path provided is a JSON, it instantiates a juggler.DataSource with the config as the only argument
+ * Else it requires it like a compiled loopback datasource
+ * @param path
+ * @returns juggler.DataSource
+ */
+function loadDataSource(path) {
+  const ds = require(path);
+  const key = Object.keys(ds)[0];
+  const val = new ds[key]();
+  return val;
+}
+
+/**
+ * Loads a compiled loopback datasource by name
+ * @param name {string}
+ * @returns {*}
+ */
+function loadDataSourceByName(name) {
+  debug(`Searching for specified dataSource ${name}`);
+  const dataSourceFiles = getAllDataSourceFiles();
+  debug(`Loaded ${dataSourceFiles.length} dataSource files`);
+
+  for (let i = 0; i < dataSourceFiles.length; i++) {
+    const f = dataSourceFiles[i];
+    const ds = loadDataSource(path.resolve(DEFAULT_DATASOURCE_DIRECTORY, f));
+    if (ds.name === name) {
+      debug(`Found dataSource ${name}`);
+      return ds;
+    } else {
+      debug(`Did not match dataSource ${name} !== ${ds.name}`);
+    }
+  }
+  throw new Error(
+    `Cannot find datasource "${name}" in ${DEFAULT_DATASOURCE_DIRECTORY}`,
+  );
+}
+
+const DEFAULT_DATASOURCE_DIRECTORY = './dist/src/datasources';
+
+const MODEL_TEMPLATE_PATH = path.resolve(
+  __dirname,
+  '../generators/model/templates/model.ts.ejs',
+);
+
+const sanitizeProperty = function(o) {
+  Object.entries(o).forEach(([k, v]) => {
+    // Delete the null properties so the template doesn't spit out `key: ;`
+    if (v === null) {
+      delete o[k];
+    }
+
+    // If you are an object or array, stringify so you don't appear as [object [object]
+    if (v === Object(v)) {
+      o[k] = JSON.stringify(o[k]);
+    }
+  });
+  // o.id could be 1 or 0, typescript will be angry if it is not a boolean
+  if (o.id) {
+    o.id = !!o.id;
+  }
+  o.tsType = o.type;
+};
+
+function getAllDataSourceFiles(dir = DEFAULT_DATASOURCE_DIRECTORY) {
```

Comment:
Here is an existing implementation for looking up all datasources. Please re-use the same low-level building blocks instead of creating new ones.

https://github.com/strongloop/loopback-next/blob/71292e9ac1b3d89ebfe284a659264cbb83dbe814/packages/cli/generators/repository/index.js#L232-L310

I think it would be great to extract `promptDataSourceName` generator action into a shared helper that can be called from multiple generators. 

Example usage of such helper in `lb4 datasource`:

```ts
const filter = item => {
    debug(`data source inspecting item: ${item}`);
    const result = utils.isConnectorOfType(
      VALID_CONNECTORS_FOR_REPOSITORY,
      this.artifactInfo.datasourcesDir,
      item,
    );
    return result !== false;
  });
};

return promptDataSourceName(this.options.datasource, this.artifactInfo, filter)
  .then({dataSourceClass} => {
    Object.assign(this.artifactInfo, {dataSourceClass});
    })
  .catch(err => {
    debug(`Error during datasource prompt: ${err}`);
    return this.exit(err);
  });
```

Reply:
> I think it would be great to extract promptDataSourceName generator action into a shared helper that can be called from multiple generators.

Feel free to leave out this refactoring out of scope of this pull request if it's easier that way.

Reply:
I'll do this in another PR

Reply:
> Here is an existing implementation for looking up all datasources. Please re-use the same low-level building blocks instead of creating new ones.

Fixed, I use [getArtifactList](https://github.com/strongloop/loopback-next/blob/master/packages/cli/lib/utils.js#L211) now.  I'll make the general prompt in another PR

Reply:
I am afraid you addressed my comment partially. It's good to see that generators are using the shared `getArtifactList` function now, but we need to change other places like `loadDataSourceByName` too, so that `getAllDataSourceFiles` and `DEFAULT_DATASOURCE_DIRECTORY` can be removed completely.

My concern is that right now, the convention for layout of datasource files is duplicated in multiple places and when we change it in the future, we won't update all of these places. 

Anyhow, I am ok to leave further improvements in this area out of scope of this initial pull request.